### PR TITLE
Switch over to rapids-cmake patches for thrust

### PR DESCRIPTION
**Pull request by robertmaynard**
## Description
 Now that rapids-cmake supports custom patches we can move cudf over to rapids-cmake for Thrust. This removes the need for custom install rules in cudf for Thrust, as rapids-cmake does that for us.
 
 This also separates out all Thrust patches so that we can better track upstream approval and remove as needed.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

**Comment by robertmaynard** [comment]
Currently blocked by https://github.com/rapidsai/rapids-cmake/pull/291 and https://github.com/rapidsai/rapids-cmake/pull/292

**Comment by robertmaynard** [comment]
Manually verified that this change hasn't regressed libcudf binary size ( 262MB for single arch ) and compile times are within a couple of seconds of branch-22.12.

```
branch-22.12  wall clock 12m 37s
this pr       wall clock 12m 49s
```

File: cpp/cmake/thirdparty/patches/thrust_override.json [hunk 1]
```diff
@@ -0,0 +1,34 @@
+
+{
+  "packages" : {
+    "Thrust" : {
+      "patches" : [
+        {
+          "file" : "Thrust/install_rules.diff",
+          "issue" : "Thrust 1.X installs incorrect files [https://github.com/NVIDIA/thrust/issues/1790]",
+          "fixed_in" : "2.0"
```
Comment (bdice) [inline cpp/cmake/thirdparty/patches/thrust_override.json:9]: Should these versions be like `X.Y.Z`? Applies several places.
```suggestion
          "fixed_in" : "2.0.0"
```

File: cpp/cmake/thirdparty/patches/thrust_override.json [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+        },
+        {
+          "file" : "${current_json_dir}/thrust_transform_iter_with_reduce_by_key.diff",
+          "issue" : "Support transform iterator with reduce by key [https://github.com/NVIDIA/thrust/pull/1805]",
+          "fixed_in" : "2.1"
+        },
+        {
+          "file" : "${current_json_dir}/thrust_disable_64bit_dispatching.diff",
+          "issue" : "Remove 64bit dispatching as not needed by libcudf [https://github.com/rapidsai/cudf/pull/11437]",
```
Comment (bdice) [inline cpp/cmake/thirdparty/patches/thrust_override.json:18]: This should mention the actual problem (compile time).
```suggestion
          "issue" : "Remove 64bit dispatching as not needed by libcudf and results in compiling twice as many kernels [https://github.com/rapidsai/cudf/pull/11437]",
```

File: cpp/cmake/thirdparty/patches/thrust_override.json [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+          "fixed_in" : ""
+        },
+        {
+          "file" : "${current_json_dir}/thrust_faster_sort_compile_times.diff",
+          "issue" : "Improve Thrust sort compile times [https://github.com/rapidsai/cudf/pull/10577]",
```
Comment (bdice) [inline cpp/cmake/thirdparty/patches/thrust_override.json:23]:
```suggestion
          "issue" : "Improve Thrust sort compile times by not unrolling loops [https://github.com/rapidsai/cudf/pull/10577]",
```

File: cpp/cmake/thirdparty/patches/thrust_override.json [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+          "fixed_in" : ""
+        },
+        {
+          "file" : "${current_json_dir}/thrust_faster_scan_compile_times.diff",
+          "issue" : "Improve Thrust scane compile times [https://github.com/rapidsai/cudf/pull/8183]",
```
Comment (bdice) [inline cpp/cmake/thirdparty/patches/thrust_override.json:28]: I don't know what the description for this one should be. Something with policy versions. 🤷‍♂️
```suggestion
          "issue" : "Improve Thrust scan compile times [https://github.com/rapidsai/cudf/pull/8183]",
```

**Review by bdice — COMMENTED** [review]
I _love_ this change. It will be so much easier to test and update our patches in the future.

File: cpp/cmake/thirdparty/patches/thrust_override.json (hunk 3 above)
Comment (davidwendt) [reply cpp/cmake/thirdparty/patches/thrust_override.json:23]: Specifically we are disabling loop-unrolling only where the comparator is in-lined. We do not disable loop unrolling for just copying data values for example.

File: cpp/cmake/thirdparty/patches/thrust_override.json (hunk 4 above)
Comment (davidwendt) [reply cpp/cmake/thirdparty/patches/thrust_override.json:28]: Like the dispatch change, this one also helps reduce the number of kernels generated for `scan` and `reduce`. @elstehle

File: cpp/cmake/thirdparty/patches/thrust_override.json [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+    "Thrust" : {
+      "patches" : [
+        {
+          "file" : "Thrust/install_rules.diff",
+          "issue" : "Thrust 1.X installs incorrect files [https://github.com/NVIDIA/thrust/issues/1790]",
+          "fixed_in" : "2.0"
+        },
+        {
+          "file" : "${current_json_dir}/thrust_transform_iter_with_reduce_by_key.diff",
+          "issue" : "Support transform iterator with reduce by key [https://github.com/NVIDIA/thrust/pull/1805]",
```
Comment (harrism) [inline cpp/cmake/thirdparty/patches/thrust_override.json:13]:
```suggestion
          "issue" : "Support transform_output_iterator as output of reduce by key [https://github.com/NVIDIA/thrust/pull/1805]",
```

File: cpp/cmake/thirdparty/patches/thrust_override.json (hunk 3 above)
Comment (bdice) [reply cpp/cmake/thirdparty/patches/thrust_override.json:23]: I edited my suggestion above.

File: cpp/cmake/thirdparty/patches/thrust_override.json (hunk 1 above)
Comment (robertmaynard) [reply cpp/cmake/thirdparty/patches/thrust_override.json:9]: Techincall it doesn't matter as the CMake `VERSION_LESS` check will consider `2`, `2.0`, and `2.0.0` all to be the 'same'. But for consistentcy across rapids-cmake we should use `2.0.0`.

File: cpp/cmake/thirdparty/patches/thrust_override.json (hunk 2 above)
Comment (robertmaynard) [reply cpp/cmake/thirdparty/patches/thrust_override.json:18]: Good point

**Review by bdice — APPROVED** [review]
Awesome. Thanks.

File: cpp/cmake/thirdparty/patches/thrust_override.json (hunk 4 above)
Comment (robertmaynard) [reply cpp/cmake/thirdparty/patches/thrust_override.json]: I updated the issue string to a better explanation ( and fixed the typo ).

**Review by davidwendt — APPROVED** [review]

**Comment by robertmaynard** [comment]
@gpucibot merge

**Pull request merged**